eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/CertSpec.hs

```diff
@@ -0,0 +1,193 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Test.Cardano.Ledger.Conway.Imp.CertSpec where
+
+import Cardano.Ledger.Address
+import Cardano.Ledger.Alonzo.Plutus.Context (ContextError)
+import Cardano.Ledger.Alonzo.Plutus.Evaluate (CollectError (BadTranslation))
+import Cardano.Ledger.Alonzo.Rules (AlonzoUtxosPredFailure (CollectErrors))
+import Cardano.Ledger.BaseTypes
+import Cardano.Ledger.Coin
+import Cardano.Ledger.Conway.Core
+import Cardano.Ledger.Conway.Governance
+import Cardano.Ledger.Conway.TxInfo (ConwayContextError (..))
+import Cardano.Ledger.Credential
+import Cardano.Ledger.Plutus.Data (Data (Data), hashData)
+import Cardano.Ledger.Plutus.Language (Plutus, PlutusLanguage)
+import Cardano.Ledger.Shelley.LedgerState
+import Data.Default.Class
+import qualified Data.Map.Strict as Map
+import qualified Data.OSet.Strict as OSet
+import qualified Data.Sequence.Strict as SSeq
+import qualified Data.Set as Set
+import Lens.Micro
+import qualified PlutusLedgerApi.V1 as P1
+import Test.Cardano.Ledger.Conway.ImpTest
+import Test.Cardano.Ledger.Core.Rational
+import Test.Cardano.Ledger.Core.Utils
+import Test.Cardano.Ledger.Imp.Common
+import Test.Cardano.Ledger.Plutus (ScriptTestContext (..))
+import Test.Cardano.Ledger.Plutus.Examples (guessTheNumber2, guessTheNumber2V2)
+
+spendDatum :: P1.Data
+spendDatum = P1.I 2
+
+setupScriptPlutus ::
+  (AlonzoEraScript era, PlutusLanguage p) =>
+  Plutus p ->
+  ImpTestM era (ScriptHash (EraCrypto era))
+setupScriptPlutus script =
+  impAddPlutusScript
+    ScriptTestContext
+      { stcScript = script
+      , stcArgs =
+          PlutusArgs
+            { paSpendDatum = Just spendDatum
+            , paData = spendDatum
+            }
+      }
+
+setupScriptPlutusV1 :: AlonzoEraScript era => ImpTestM era (ScriptHash (EraCrypto era))
+setupScriptPlutusV1 = setupScriptPlutus guessTheNumber2
+
+setupScriptPlutusV2 :: AlonzoEraScript era => ImpTestM era (ScriptHash (EraCrypto era))
+setupScriptPlutusV2 = setupScriptPlutus guessTheNumber2V2
+
+txWithPlutus ::
+  forall era.
+  ConwayEraImp era =>
+  ImpTestM era (ScriptHash (EraCrypto era)) ->
+  ImpTestM era (Tx era)
+txWithPlutus setupScript = do
+  sh <- setupScript
+  submitTxAnn "Submit a Plutus" $
+    mkBasicTx mkBasicTxBody
+      & bodyTxL . outputsTxBodyL
+        .~ SSeq.singleton
+          ( mkBasicTxOut
+              (Addr Testnet (ScriptHashObj sh) StakeRefNull)
+              (inject (Coin 100))
+              & dataHashTxOutL .~ SJust (hashData @era $ Data spendDatum)
+          )
+
+testPlutusV1V2Failure ::
+  forall era a.
+  ( ConwayEraImp era
+  , InjectRuleFailure "LEDGER" AlonzoUtxosPredFailure era
```

Comment:
All helper functions that can result in a test failure must have `HasCallStack` constraint, otherwise you won't know which occurance is at fault:
```suggestion
  , InjectRuleFailure "LEDGER" AlonzoUtxosPredFailure era
  , HasCallStack
```